Kiba 2024 Migrator

awisema opened the pull request.

awisema committed:
2024 HUD Data Models.

awisema committed:
WIP: CSV Migrator
Exporter
Project Descriptors

awisema committed:
WIP: CSV Migrator
Client

awisema committed:
WIP: CSV Migrator
Enrollment

awisema committed:
WIP: DB Migrator

awisema committed:
WIP: DB Migrator
Bugfixes

awisema committed:
DB Migrator fixes
HUD 'I' values are numbers from the DB, but strings from the CSV.
Data_source_id is required when inserting DB records.

eanders commented on app/models/grda_warehouse/hud/ce_participation.rb:
```diff
@@ -0,0 +1,33 @@
+    belongs_to :data_source, optional: true
+
+    # hide previous declaration of :importable, we'll use this one
+    replace_scope :importable, -> do
+      where(synthetic: false)
+    end
+
+    scope :synthetic, -> do
+      where(synthetic: true)
+    end
```
This table doesn't have a `synthetic` column, pretty sure it shouldn't have these scopes. ^^^

eanders commented on app/models/grda_warehouse/hud/hmis_participation.rb:
```diff
@@ -0,0 +1,33 @@
+    belongs_to :project, **hud_assoc(:ProjectID, 'Project'), inverse_of: :hmis_participations, optional: true
+    belongs_to :data_source, optional: true
+
+    # hide previous declaration of :importable, we'll use this one
+    replace_scope :importable, -> do
+      where(synthetic: false)
+    end
+
+    scope :synthetic, -> do
+      where(synthetic: true)
+    end
```
synthetic isn't a thing in this table.
```suggestion
    # hide previous declaration of :importable, we'll use this one
    replace_scope :importable, -> do
      where(synthetic: false)
    end

    scope :synthetic, -> do
      where(synthetic: true)
    end
```

eanders commented on db/health_structure.sql:
Pretty sure this file shouldn't have changed.

eanders commented on db/reporting_structure.sql:
This file shouldn't be changed either.

eanders commented on db/structure.sql:
Also, no change here right?

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/concerns/hud_twenty_twenty_two_to_twenty_twenty_four/references.rb:
```diff
@@ -0,0 +1,34 @@
+###
+
+module HudTwentyTwentyTwoToTwentyTwentyFour
+  module References
```
Should this be something like CsvReferences?  I think you mentioned you'd need DB references as well, but this one reads a CSV file, which feels like it's probably not reusable for the db.

eanders replied:
Oh, maybe that's the model vs file below.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb:
```diff
@@ -0,0 +1,28 @@
+        reference(:enrollment_coc) do |row|
+          next unless row['DataCollectionStage'].to_i == 1
+
+          h[row['HouseholdID']] ||= row['CoCCode']
```
I think we'll need to handle empty HouseholdID, if we don't, everyone with a nil HouseholdID will end up with the same CoCCode.  Not sure if we should use EnrollmentID as a backup, or just return a blank CoCCode.  Open to suggestions.

awisema replied:
Good point. This seems like a recipe for making bad DQ even worse...
The 2022 spec says:

> There must be one and only one record in EnrollmentCoC.csv with a DataCollectionStage of 1 for each EnrollmentID in Enrollment.csv where RelationshipToHoH is equal to 1. Additionally, there must be one and only one record in EnrollmentCoC.csv with a DataCollectionStage of 1 for each HouseholdID in Enrollment.csv.

So, I don't think using the EnrollmentID will help much. and I don't think leaving it blank would be making the DQ any worse than it already is.

Alternatively, the 2024 spec for the CoCCode column says:

> Must match a record in ProjectCoC.csv with the same ProjectID

So, we could use the code from ProjectCoC when there is only a single CoC Code for the Project given that_most_ projects have only one, if you think it's worth it.

awisema committed:
Merge branch 'release-86' into aw/kiba-2024-migrator-185275486

# Conflicts:
#	db/reporting_structure.sql
#	db/structure.sql
#	db/warehouse_structure.sql

awisema committed:
Review cleanup
Participation tables don't have synthetic options.

eanders commented on db/reporting_structure.sql:
This one's still here.

eanders commented on db/structure.sql:
This one too.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb:
```diff
@@ -0,0 +1,28 @@
+      row
+    end
+
+    private def entry_enrollment_coc(household_id)
+      @entry_enrollment_coc ||= {}.tap do |h|
+        reference(:enrollment_coc) do |row|
+          next unless row['DataCollectionStage'].to_i == 1
+
+          h[row['HouseholdID']] ||= row['CoCCode']
+        end
+      end
+      @entry_enrollment_coc[household_id]
+    end
```
```suggestion
    private def entry_enrollment_coc(enrollment_row)
      @entry_enrollment_coc ||= {}.tap do |h|
        reference(:enrollment_coc) do |row|
          next unless row['DataCollectionStage'].to_i == 1

          key = row['HouseholdID'].presence || "en_#{row['EnrollmentID']}" 
          key += "_ds_#{row['data_source_id']}"
          h[key] ||= row['CoCCode'] if HudUtility2024.valid_coc?(row['CoCCode'])
        end
      end
      key = enrollment_row['HouseholdID'].presence || "en_#{enrollment_row['EnrollmentID']}" 
      key += "_ds_#{row['data_source_id']}"
      @entry_enrollment_coc[key]
    end
```

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb:
```diff
@@ -0,0 +1,28 @@
+    include ::HudTwentyTwentyTwoToTwentyTwentyFour::References
+
+    def process(row)
+      row['EnrollmentCoC'] = entry_enrollment_coc(row['HouseholdID'])
```
```suggestion
      row['EnrollmentCoC'] = entry_enrollment_coc(row)
```

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/client/update_race_and_ethnicity.rb:
```diff
@@ -0,0 +1,19 @@
+  class UpdateRaceAndEthnicity
+    def process(row)
+      row['HispanicLatinaeo'] = row['Ethnicity'] == 1 ? 1 : 0 # 8/9/99 all convert to 0
+      # row['RaceNone'] = nil if row['HispanicLatinaeo'] == 1 # Force clearing of RaceNone if we set HispanicLatinaeo
```
Was this intentionally left commented?

awisema replied:
Oops, no! This was from when I was avoiding stepping on my DB for testing.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/csv_transformer.rb:
```diff
@@ -0,0 +1,230 @@
+      end
+    end
+
+    def self.fix_bad_line_endings(filename, encoding)
+      tmp_file = ::Tempfile.new(filename)
+      file_with_bad_line_endings = false
+
+      File.open(filename, 'r', encoding: encoding) do |file|
+        file_with_bad_line_endings = ! valid_line_endings?(file)
+      end
+
+      if file_with_bad_line_endings
+        File.open(filename, 'r', encoding: encoding) do |file|
+          copy_length = file.stat.size - 2
+          Rails.logger.debug "Correcting bad line ending in #{filename}"
+          File.copy_stream(file, tmp_file, copy_length, 0)
+          tmp_file.write("\n")
+          tmp_file.close
+        end
+        FileUtils.cp(tmp_file, filename)
+      end
+    ensure
+      tmp_file&.close
+      tmp_file&.unlink
+    end
```
What do you think about moving this logic up in to `/app/models/something` or maybe as a monkey patch for CSV?  I feel like we call this pretty often, might be nice to centralize it.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb:
```diff
@@ -0,0 +1,64 @@
+
+      15 => 215,
+      6 => 206,
+      7 => 207,
+      25 => 225,
+      4 => 204,
+      5 => 205,
+
+      2 => 302,
+      29 => 329,
+      14 => 314,
+      32 => 332,
+      12 => 312,
+      13 => 313,
+      27 => 227,
+      36 => 336,
+      35 => 335,
+
+      22 => 422,
+      23 => 423,
+      26 => 426,
+      10 => 410,
+      21 => 421,
+      11 => 411,
+
+      30 => 30,
+      17 => 17,
+      24 => 24,
+      37 => 37,
+      8 => 8,
+      9 => 9,
+      99 => 99,
+
+      28 => 435,
+      19 => 435,
+      31 => 435,
+      34 => 435,
+      20 => 435,
+      3 => 435,
+    }.freeze
```
These appear to be duplicative of `drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/change_living_situation.rb` Can we centralize that? Maybe in the structure file?

eanders replied:
Actually in HudUtility2024 probably makes more sense.  Something like `def cls_migration_for_2022_to_2024`

awisema replied:
I moved it to a concern in the migrator -- We shouldn't need it outside of migration, so HUDUtility seems like the wrong place to me.

eanders replied:
Thanks

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/add_rental_subsidy_type.rb:
```diff
@@ -0,0 +1,29 @@
+    def process(row)
+      row['RentalSubsidyType'] = nil
+
+      situation = row['LivingSituation'].to_i
```
is `LivingSituation` the right thing to check here?

awisema replied:
Yes -- the mapping document is looking for 3.917A.1 or 3.917B.1 both of which are stored in `LivingSituation` on Enrollment.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/change_living_situation.rb:
```diff
@@ -0,0 +1,64 @@
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::Enrollment
+  class ChangeLivingSituation
+    LIVING_SITUATIONS = {
```
Another instance we should DRY

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/move_enrollment_coc.rb:
```diff
@@ -0,0 +1,28 @@
+      row
+    end
+
+    private def entry_enrollment_coc(household_id)
+      @entry_enrollment_coc ||= {}.tap do |h|
+        reference(:enrollment_coc) do |row|
+          next unless row['DataCollectionStage'].to_i == 1
+
+          h[row['HouseholdID']] ||= row['CoCCode']
+        end
+      end
+      @entry_enrollment_coc[household_id]
+    end
```
See the other proposed fix.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/exit/add_destination_subsidy_type.rb:
```diff
@@ -0,0 +1,30 @@
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::Exit
+  class AddDestinationSubsidyType
+    SUBSIDY_TYPES = {
+      28 => 428,
+      19 => 419,
+      31 => 431,
+      34 => 433,
+      20 => 434,
+      3 => 420,
+    }.freeze
```
DRY?

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/exit/update_destination.rb:
```diff
@@ -0,0 +1,64 @@
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::Exit
+  class UpdateDestination
+    DESTINATIONS = {
```
These are probably the same as the living situations.

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/hmis_participation/create_hmis_participation.rb:
```diff
@@ -0,0 +1,52 @@
+    private def victim_service_providers
+      @victim_service_providers ||= {}.tap do |h|
+        reference(:organization) do |row|
+          h[row['OrganizationID']] = row['VictimServiceProvider']
```
Need `data_source_id` for the DB version

eanders commented on drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/project/update_project_type.rb:
```diff
@@ -0,0 +1,25 @@
+module HudTwentyTwentyTwoToTwentyTwentyFour::Project
+  class UpdateProjectType
+    def process(row)
+      project_type = if row['ProjectType'].to_i == 1
+        if row['TrackingMethod'].to_i == 3
+          1
+        else
+          0
+        end
+      else
+        row['ProjectType']
+      end
```
Nothing wrong with what you have, but maybe:
```suggestion
      project_type = if row['ProjectType'].to_i == 1 && row['TrackingMethod'].to_i == 3
          0
      else
        row['ProjectType']
      end
```

awisema replied:
I'm going to stick w/ mine... I think yours needed to be `row['TrackingMethod'].to_i == 0` which is fine, but if we both had to work through it to know if it is correct...

eanders reviewed (changes requested):
A few changes in-line.  Also, maybe we could add some really simple tests where we run a 2022 CSV and convert it to a 2024 and confirm the results match?   Even if we generate the results using the migrator, it would confirm it wasn't throwing an error, and we could eyeball the result file manually to confirm things are processed as expected.

awisema committed:
DRY up 3.12.1 Living Situation Options and 3.12.A RentalSubsidyTypes

awisema committed:
Use EnrollmentID + data_source_id if HouseholdID is blank

Co-authored-by: Elliot <elliot@greenriver.org>

awisema committed:
Uncomment code that was commented for testing.

awisema committed:
Include data_source_id when processing DB.

awisema committed:
Correct generated timestamp format
To do a transform to CSV, generate a correct datetime string in the HMISParticpation generator.
To do a transform in DB, generate a Time object.

awisema committed:
Simple smoketest.

awisema committed:
Move input fixture to S3

awisema committed:
Add moved test input to .gitignore

awisema committed:
Pull accidentially touched structure files from branchpoint

awisema committed:
Move fix_bad_line_endings to an app concern to reduce code duplication.

awisema committed:
Merge branch 'release-86' into aw/kiba-2024-migrator-185275486

awisema committed:
Merge branch 'release-86' into aw/kiba-2024-migrator-185275486

# Conflicts:
#	db/warehouse_structure.sql

awisema committed:
Merge branch 'release-86' into aw/kiba-2024-migrator-185275486

eanders commented on .github/workflows/rails_tests.yml:
```diff
@@ -193,6 +193,17 @@ jobs:
+        uses: keithweaver/aws-s3-github-action@v1.0.0
+        with:
+          command: cp
+          source: ${{ secrets.AWS_S3_ACTION_BUCKET }}/${{ vars.AWS_S3_ACTIONS_SOURCE_PATH_4_2023 }}
```
I'm guessing you need 
```suggestion
          source: ${{ secrets.AWS_S3_ACTION_BUCKET }}${{ vars.AWS_S3_ACTIONS_SOURCE_PATH_4_2023 }}
```

awisema committed:
Update .github/workflows/rails_tests.yml

Co-authored-by: Elliot <elliot@greenriver.org>

gigxz commented:
Looks like there are some test failures in `hud_twenty_twenty_two_to_twenty_twenty_four` fyi

awisema committed:
BUGFIX: Missing directories on test path

gigxz commented:
@awisema tests still slightly unhappy 🧐 
would you mind also merging in the latest 86, so I can get a reasonable diff on https://github.com/greenriver/hmis-warehouse/pull/3383? I'm waiting to request review since it's difficult to review in the current state

awisema committed:
BUGFIX: Adjust test paths to match S3

awisema committed:
Merge branch 'release-86' into aw/kiba-2024-migrator-185275486

# Conflicts:
#	.gitignore

awisema committed:
More test paths

awisema committed:
Re-enable tests.
